README.md

[![Version](https://img.shields.io/cocoapods/v/AlertWizard.svg?style=flat)](http://cocoapods.org/pods/AlertWizard)
[![License](https://img.shields.io/cocoapods/l/AlertWizard.svg?style=flat)](http://cocoapods.org/pods/AlertWizard)
[![Platform](https://img.shields.io/cocoapods/p/AlertWizard.svg?style=flat)](http://cocoadocs.org/docsets/AlertWizard)
[![Platform](https://img.shields.io/badge/Swift-5-brightgreen.svg?style=flat)](http://cocoapods.org/pods/AlertWizard)


## Overview

# AlertWizard

AlertWizard provides centralized control and structure to alert messages. It uses UIAlertController in a way intended by Apple, while providing compact and transparent methods that are easy to use, as well as framework for managing your own alert messages in an elegant way. 

You can create your own fully custom alert controllers implementing AlertDisplayer protocol, and have them displayed via AlertWizard.  

Store all your alert messages, headers, buttons, ect in an easily localizable .json file

## Installation

AlertWizard is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "AlertWizard"
```

To generate an empty Podfile, navigate to your project folder and execute the following command:

```shell
$ pod init
```

If you do not have CocoaPods installed, run installation command first:

```shell
$ sudo gem install cocoapods
```

If you do not want to use CocoaPods, you can checkout the repository and copy framework files to your project manually.

## Requirements

- iOS11 and higher

## Author

Lesya V, lveecode@gmail.com

## License

Available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Installation & Update

To install or update CocoaPods see this [guide](http://docs.cocoapods.org/guides/installing_cocoapods.html).

To update AlertWizard use:

```shell
$ pod update AlertWizard
```

[//]: # (## 0.1.0.1)

[//]: # (##### Enhancements)

[//]: # (* Enhancement)

[//]: # (##### Bug Fixes)

[//]: # (* Bug fix)


Example/Tests/Tests.swift
import XCTest
import AlertWizard

class Tests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


Example/AlertWizard/ViewController.swift
//
//  ViewController.swift
//  AlertWizard
//
//  Created by LVeecode on 07/10/2019.
//  Copyright (c) 2019 LVeecode. All rights reserved.
//

import UIKit
import AlertWizard

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    //
    // MARK: - Actions
    //
    @IBAction func alertButtonAction() {
        AlertHelper.displayAlertController(forReason: AlertCodes.TestAlert.rawValue, delegate: self)
    }

}

extension ViewController: AlertDisplayerDelegate {
    
    func didUseAction(atIndex index: Int, alertDisplayer displayer: AlertDisplayer, reason: Int) {
        print("action at index ", index)
        print("text field content: ", displayer.textFields?.first?.text ?? "none")
    }
    
    func didUseCancelActionOfAlert(withReason reason: Int) {
        print("cancel action")
    }
    
    func didUseDestructiveActionOfAlert(withReason reason: Int) {
        print("destructive action")
    }
    
}



AlertWizard/Classes/AlertController.swift
//
//  AlertController.swift
//  CulinarySubs
//
//  Created by Lesya Verbina on 6/26/19.
//  Copyright © 2019 Lesya Verbina. All rights reserved.
//

import UIKit

class AlertController: UIAlertController, AlertDisplayer {
    
    var alertWindow: UIWindow? = nil
    var tintColor: UIColor? = nil
    var statusBarHidden: Bool = false
    var statusBarStyle: UIStatusBarStyle = .default
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        // Adding fade-out
        UIView.animate(withDuration: 0.1) { [weak self] in
            self?.alertWindow?.alpha = 0
        }
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        
        // Precaution to ensure window gets destroyed
        alertWindow?.isHidden = true
        alertWindow = nil
    }
    
    func show(animated: Bool, blur: Bool) {
        
        // Save topmost controller's status bar style
        
        let mainWindow: UIWindow = UIApplication.shared.keyWindow!
        
        var responsibleController: UIViewController? = mainWindow.rootViewController?.presentedViewController != nil ? mainWindow.rootViewController?.presentedViewController : mainWindow.rootViewController
        
        if responsibleController?.isKind(of: UITabBarController.self) ?? false {
            
            responsibleController = (responsibleController as! UITabBarController).selectedViewController
        }
        
        if responsibleController?.isKind(of: UINavigationController.self) ?? false {
            
            let responsibleNavController = responsibleController as! UINavigationController
            
            if responsibleNavController.isNavigationBarHidden ||
                responsibleNavController.navigationBar.isTranslucent {
                
                responsibleController = responsibleNavController.topViewController
            }
        }
        
        if responsibleController == nil { responsibleController = UIViewController() }
        
        statusBarStyle = responsibleController!.preferredStatusBarStyle
        statusBarHidden = responsibleController!.prefersStatusBarHidden
        
        alertWindow = UIWindow(frame: UIScreen.main.bounds)
        
        // Force status bar to have particular style via UINavigationController
        let rootNavController = UINavigationController()
        rootNavController.navigationBar.barStyle = statusBarStyle == .default ? UIBarStyle.default : UIBarStyle.black
        rootNavController.isNavigationBarHidden = true
        
        alertWindow?.rootViewController = rootNavController
        
        // We inherit the main window's tintColor
        alertWindow?.tintColor = UIApplication.shared.delegate?.window??.tintColor
        // Window level is above the top window (this makes the alert, if it's a sheet, show over the keyboard)
        let topWindow = UIApplication.shared.windows.last
        alertWindow?.windowLevel = topWindow?.windowLevel ?? UIWindow.Level.alert + 1
        
        if blur {
            // Create blur effect
            let blurEffect = UIBlurEffect(style: .light)
            // Add effect to an effect view
            let visualEffectView = UIVisualEffectView(effect: blurEffect)
            visualEffectView.frame = alertWindow?.frame ?? UIScreen.main.bounds
            alertWindow?.addSubview(visualEffectView)
        }
        
        alertWindow?.makeKeyAndVisible()
        
        // Animate fade in
        alertWindow?.alpha = 0
        UIView.animate(withDuration: 0.2) { [weak self] in
            self?.alertWindow?.alpha = 1
        }
        
        // Needed for iPad presentation, will present from the bottom center of the screen
        popoverPresentationController?.sourceView = alertWindow?.rootViewController?.view
        popoverPresentationController?.sourceRect = CGRect(
            x: alertWindow?.rootViewController?.view.bounds.size.width ?? 0 / 2,
            y: alertWindow?.rootViewController?.view.bounds.size.height ?? 0,
            width: 1,
            height: 1)
        
        modalTransitionStyle = .crossDissolve
        modalPresentationStyle = .fullScreen
        
        alertWindow?.rootViewController?.present(self, animated: animated, completion: nil)
        
        // Customize buttons tint color after presenting, or it won't work
        if tintColor != nil {
            alertWindow?.tintColor = tintColor
        }
    }
    
    func hide(animated: Bool) {
        // Alert controller is hidden automatically
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return statusBarStyle
    }
    
    override var prefersStatusBarHidden: Bool {
        return statusBarHidden
    }
    
    static func create(withTitle title: String?,
                       message: String?,
                       buttonCount: Int,
                       textFieldPlaceholders: [String],
                       tintColor: UIColor?) -> AlertDisplayer {
        
        // If more than 2 buttons - action sheet
        let alertController = AlertController.init(title: title,
                                                   message: message,
                                                   preferredStyle: (buttonCount > 1 && textFieldPlaceholders.count == 0) ? .actionSheet : .alert)
        
        // Custom tint color, if any
        if tintColor != nil {
            alertController.tintColor = tintColor
        }
        
        // Text fields, if any
        for textFieldPlaceholder in textFieldPlaceholders {
            alertController.addTextField { (textField) in
                textField.placeholder = textFieldPlaceholder
            }
        }
        
        return alertController
    }

    func add(buttonActionDicts: [NSDictionary],
             cancelActionDict: NSDictionary,
             destructiveActionDict: NSDictionary) {
        
        // Cancel action
        let cancelTitle: String = cancelActionDict.allKeys.first as? String ?? "Ok"
        let cancelBlock: (()->Void)? = cancelActionDict.allValues.first as? (()->Void)
        
        let cancelAction = UIAlertAction.init(title: cancelTitle,
                                              style: .cancel) { (action) in
            cancelBlock?()
        }
        addAction(cancelAction)
        
        // Action buttons, if any
        for actionDict in buttonActionDicts {
            let actionTitle: String = actionDict.allKeys.first as! String
            let actionBlock: (()->Void)? = actionDict.allValues.first as? (()->Void)
            
            let action = UIAlertAction.init(title: actionTitle, style: .default) { (action) in
                actionBlock?()
            }
            addAction(action)
        }
        
        // Destructive action, if any
        if destructiveActionDict.count > 0 {
            let destructiveTitle: String = destructiveActionDict.allKeys.first as! String
            let destrBlock: (()->Void)? = destructiveActionDict.allValues.first as? ()->Void
            
            let destrAction = UIAlertAction.init(title: destructiveTitle,
                                                 style: .destructive) { (action) in
                                                    destrBlock?()
            }
            addAction(destrAction)
        }
    }
}


AlertWizard/Classes/AlertHelper.swift
//
//  AlertHelper.swift
//  CulinarySubs
//
//  Created by Lesya Verbina on 6/26/19.
//  Copyright © 2019 Lesya Verbina. All rights reserved.
//

import UIKit

/// Displayer protocol for custom alert controllers.
@objc public protocol AlertDisplayer {
    
    /// Access the text fields.
    @objc var textFields: [UITextField]? { get }
    
    /// Initialize a displayer with alert info.
    ///
    /// - Parameters:
    ///   - title: Alert title, can be nil.
    ///   - message: Alert body, can be nil.
    ///   - buttonCount: Number of action buttons, can be used to determine the style of alert controller.
    ///   - textFieldPlaceholders: Array of text field placeholders.
    ///   - tintColor: Tint color for the buttons
    @objc static func create(withTitle title: String?,
                             message: String?,
                             buttonCount: Int,
                             textFieldPlaceholders: [String],
                             tintColor: UIColor?) -> AlertDisplayer
    
    ///  Add buttons and action blocks. Separate from initializer so that the instance can be passed on into the blocks.
    ///
    /// - Parameters:
    ///   - buttonActions: Array of dictionaries with action button titles and blocks, e.g. @{ @"Log out" : block}, @{@"Reset pin" : block }
    ///   - cancelAction: Dictionary with cancel button title and block, e.g. @{ @"Cancel" : block }
    ///   - destructiveAction: Dictionary with destructive button title and block.
    /// - Warning: Blocks are memory stack based, so you need to [block copy] before adding to the dictionary, if you want them to work as expected.
    @objc func add(buttonActionDicts: [NSDictionary],
                   cancelActionDict: NSDictionary,
                   destructiveActionDict: NSDictionary)
    
    /// Display the alert on screen.
    ///
    /// - Parameters:
    ///   - animated: Flag to animate the transition.
    ///   - blur: Flag to blur the background.
    @objc func show(animated: Bool, blur: Bool)
    
    
    /// Hide the alert
    ///
    /// - Parameter animated: Flag to animate the transition
    @objc func hide(animated: Bool)
}

/// Delegate much like UIAlertView delegate.
@objc public protocol AlertDisplayerDelegate {
    
    /// This method is called when dismiss action of an alert displayer is used.
    ///
    /// - Parameter reason: Reason of the alert displayer, as provided on creation.
    @objc optional func didUseCancelActionOfAlert(withReason reason: Int)
    
    /// This method is called when a destructive action of an alert displayer is used.
    ///
    /// - Parameter reason: Reason of the alert displayer, as provided on creation.
    @objc optional func didUseDestructiveActionOfAlert(withReason reason: Int)
    
    /// This method is called when one of the actions of displayer is used (except the cancel and destructive actions, those have their own delegate methods)
    ///
    /// - Parameters:
    ///   - index: Index of a used action.
    ///   - displayer: Alert displayer object, provided here so text fields can be accessed.
    ///   - reason: Reason of the alert displayer, as provided on creation.
    @objc func didUseAction(atIndex index: Int,
                            alertDisplayer displayer: AlertDisplayer,
                            reason: Int)
}

/**
 * Alert reasons will have to be provided as integers, and stored in enum format in your project.
 * Their ids need to correspond those in alerts JSON file, as this is the way JSON alerts will be retrieved.
 *
 * Example of alert reason enum:
 * enum {
 *      AlertReason = 100,
 *  };
 *
 * Corresponding JSON file
 * {
 *   "100": {
 *          "reason" : "AlertReason", // Readable reason, not used in code
 *          "title" : "Accept user?",   // Alert title
 *          "body" : "You can enter short message", // Alert body
 *          "action_titles" : ["Accept user", "Decline user"], // OPTIONAL Action titles,
 *          "cancel_title" : "Cancel", // OPTIONAL Cancel button title, if not provided, "OK" will be used
 *          "text_fields" : ["Message"], // OPTIONAL text fields to be added to controller, providing placeholder text as string
 *      }
 *  }
 */
@objc public class AlertHelper: NSObject {
    
    //
    // MARK: - Customization
    //
    
    ///  Set tint color of alert controller window so button's text color is customized.
    @objc public var tintColor: UIColor? = nil
    
    /// Enable or disable full screen blur when the alert is displayed.
    @objc public var blurEnabled: Bool = false
    
    /// Path for alerts file, set it here in case it is not "Alerts.json" located in the main Bundle.
    /// Used for localization
    @objc public var pathToAlertsFile: String? = Bundle.main.path(forResource: "Alerts", ofType: "json")
    
    //
    // MARK: - default UIAlertController summoning
    //
    
    /// Method that creates and displays standart alert controller with provided reason. Info for controller will be taken from Alerts.json file in main bundle.
    /// Delegate will receive callbacks on dissmiss, destructive and regular actions.
    /// Can provide arguments for formatted title or message body.
    ///
    /// - Parameters:
    ///   - reason: Reason of alert controler, by which alert details will be accessed in Alerts.json file.
    ///   - titleArguments: Arguments for formatted string in alert title.
    ///   - bodyArguments: Arguments for formatted string in alert body.
    ///   - delegate: Delegate, if any.
    /// - Returns: Alert controller object that will be shown.
    @objc @discardableResult public static func displayAlertController(forReason reason: Int,
                                                                titleArguments: [String]? = nil,
                                                                bodyArguments: [String]? = nil,
                                                                delegate: AlertDisplayerDelegate? = nil) -> UIAlertController? {
        
        return shared.displayAlertController(forReason: reason,
                                             titleArguments: titleArguments,
                                             bodyArguments: bodyArguments,
                                             delegate: delegate)
    }
    
    //
    // MARK: - Custom alert displayer summoning
    //
    
    ///  Method that creates and displays custom alert controller, which must conform to WLAlertDisplayer protocol, with provided reason.
    /// Info for controller will be taken from Alerts.json file in main bundle.
    /// Delegate will receive callbacks on dissmiss, destructive and regular actions.
    /// Can provide arguments for formatted title or message body.
    ///
    /// - Parameters:
    ///   - aClass: Class of the custom Alert Displayer.
    ///   - reason: Reason of alert displayer, by which alert details will be accessed in Alerts.json file.
    ///   - titleArguments: Arguments for formatted string in alert title.
    ///   - bodyArguments: Arguments for formatted string in alert body.
    ///   - delegate: Delegate, if any
    /// - Returns: Alert displayer object that will be shown.
    @objc @discardableResult public static func displayAlertDisplayer(ofClass aClass: AlertDisplayer.Type,
                                                               reason: Int,
                                                               titleArguments: [String]? = nil,
                                                               bodyArguments: [String]? = nil,
                                                               delegate: AlertDisplayerDelegate? = nil) -> AlertDisplayer? {

        return shared.displayAlertDisplayer(ofClass: aClass,
                                            reason: reason,
                                            titleArguments: titleArguments,
                                            bodyArguments: bodyArguments,
                                            delegate: delegate)
    }


    //
    // MARK: - Private
    //
    
    private static let shared = AlertHelper()
    private var showingAlert: Bool = false
    
    private func displayAlertController(forReason reason: Int,
                                                                titleArguments: [String]? = nil,
                                                                bodyArguments: [String]? = nil,
                                                                delegate: AlertDisplayerDelegate? = nil) -> UIAlertController? {
        
        // Returning alert controller, so that view controller
        // can access it and text fields in it, in particular
        return displayAlertDisplayer(ofClass: AlertController.self,
                                     reason: reason,
                                     titleArguments: titleArguments,
                                     bodyArguments: bodyArguments,
                                     delegate: delegate) as? UIAlertController
    }
    
    private func displayAlertDisplayer(ofClass aClass: AlertDisplayer.Type,
                                                               reason: Int,
                                                               titleArguments: [String]? = nil,
                                                               bodyArguments: [String]? = nil,
                                                               delegate: AlertDisplayerDelegate? = nil) -> AlertDisplayer? {
        
        weak var weakDelegate = delegate
        
        // If already displaying alert - not displaying new one
        if (showingAlert) {
            return nil
        }
        showingAlert = true
        
        // Parsing alert contents
        let messageDict = AlertHelper.alertDictionaryFromJSON(forReason: reason)
        var title: String? = messageDict["title"] as? String
        var message: String? = messageDict["message"] as? String
        let actionTitles: [String] = (messageDict["action_titles"] as? [String]) ?? []
        var cancelTitle: String = (messageDict["cancel_title"] as? String) ?? ""
        let destrTitle: String = (messageDict["destructive_title"] as? String) ?? ""
        let textFieldPlaceholders: [String] = (messageDict["text_fields"] as? [String]) ?? []
        if cancelTitle.count == 0 {
            cancelTitle = "Ok"
        }
        
        // If provided arguments for title formatted string of body formatted string, adding them
        if titleArguments?.count ?? 0 > 0 {
            let a = titleArguments! + ["x","x","x","x","x","x","x","x","x","x"]
            title = String(format: title ?? "", a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7], a[8], a[9], a[10])
        }
        if bodyArguments?.count ?? 0 > 0 {
            let a = bodyArguments! + ["x","x","x","x","x","x","x","x","x","x"]
            message = String(format: message ?? "", a[0], a[1], a[2], a[3], a[4], a[5], a[6], a[7], a[8], a[9], a[10])
        }
        
        // If body or title are absent, hide them so no empty
        // lines show up
        if (message?.count == 0) {
            message = nil
        }
        if (title?.count == 0) {
            title = nil
        }
        
        let displayer = aClass.create(withTitle: title,
                                      message: message,
                                      buttonCount: actionTitles.count,
                                      textFieldPlaceholders: textFieldPlaceholders,
                                      tintColor: tintColor)
        
        // Construct dictionaries for alert displayer
        let cancelBlock: (()->Void) = { [weak self] in
            displayer.hide(animated: true)
            
            self?.showingAlert = false
            weakDelegate?.didUseCancelActionOfAlert?(withReason: reason)
        }
        let cancelDict = [cancelTitle : cancelBlock]
        
        var buttonActionDicts: [NSDictionary] = []
        for (i, actionTitle) in actionTitles.enumerated() {
            let actionBlock: (()->Void) = { [weak self] in
                displayer.hide(animated: true)
                
                self?.showingAlert = false
                weakDelegate?.didUseAction(atIndex: i, alertDisplayer: displayer, reason: reason)
            }
            
            buttonActionDicts.append([actionTitle : actionBlock] as NSDictionary)
        }
        
        var destrDict: [String: Any] = [:]
        if destrTitle.count > 0 {
            let destrBlock: ()->Void = { [weak self] in
                displayer.hide(animated: true)
                
                self?.showingAlert = false
                weakDelegate?.didUseDestructiveActionOfAlert?(withReason: reason)
            }
            
            destrDict.updateValue(destrBlock, forKey: destrTitle)
        }
        
        displayer.add(buttonActionDicts: buttonActionDicts,
                      cancelActionDict: cancelDict as NSDictionary,
                      destructiveActionDict: destrDict as NSDictionary)
        
        displayer.show(animated: true, blur: blurEnabled)
        
        return displayer
    }
    
    static func alertDictionaryFromJSON(forReason reason: Int) -> Dictionary<String, Any> {
        
        if shared.pathToAlertsFile == nil { return [:] }
        let jsonDictionary: Dictionary<String, Any> = objectFromJSONFile(atPath: AlertHelper.shared.pathToAlertsFile!) as! Dictionary<String, Any>
        
        let reasonString = String(format: "%d", reason)
        return jsonDictionary[reasonString] as! Dictionary<String, Any>
    }
    
    static func objectFromJSONFile(atPath path: String) -> Any {
        
        // Parsing JSON file into array/dict
        let data: Data? = NSData.init(contentsOfFile: path) as Data?
        if data == nil { return [] }
        
        do {
            let jsonObject = try JSONSerialization.jsonObject(with: data!, options: JSONSerialization.ReadingOptions(rawValue: 0))
            return jsonObject
        }
        catch let error as NSError {
            print("Found an error - \(error)")
            return []
        }
    }
}
